### DISCUSSION TITLE
Is it possible to save the resulting page as a static .html file?

### DISCUSSION THREAD

**Discussion by smojef** [zauberzeug/nicegui#313] [Q&A] [answered]
I have a nicegui script that prompts the user and then generates a report with tables, charts, and other ui elements.  
Before shutting down my script, is it possible to save the resulting page (or at least part of it) as a static .html file for later viewing?
The static .html file should not attempt to connect to my app, but could load public resources such as quasar or tailwind for proper display.
Obviously I could ask the user to right-click the page in his browser and select Save As..., but I want to control the file name of the saved file.

**Comment by falkoschindler** [comment]
This is an interesting idea! A general solution, however, might turn out pretty complicated, since we basically need to write a dedicated renderer.
If you want to provide an export feature for a generated report, it's probably easiest to write a custom FastAPI endpoint yourself which provides exactly what is needed.
An alternative might be to use Selenium on the server. You could point it to the page you want to render and grab the generated HTML.

**Reply by smojef** [reply]
I would need to save the page in the state that it currently is following user interaction.  So the Selenium option would not work.
At least if there was a way to grab the entire HTML structure from behind?  Something we could call like `app.current_page.get_html()`?  I'm sure I could then find a hack to disable scripts by messing with the HTML structure ;-)

**Comment by Diegiwg** [comment, marked as answer]
EDIT: The function below returns the current html of the page, however, it would be the same as ``"Right Click > Save As > Web Page (HTML only)"``, that is, it will not save the **CSS** and **JS** files, which **NiceGUI** generates/uses, available in the ``_static`` folder.

So, would have to look for another solution, maybe use some JS lib to transform the page into PDF.

CODE:

```python
from nicegui import ui

async def save():
    html = await ui.run_javascript("document.querySelector('html').outerHTML")

    with open("./test.html", "w", encoding="utf-8") as file:
        file.write(html)
    
ui.button("Click", on_click=save)

ui.run()
```

-- FIRST COMMENT --

You can use ``ui.run_javascript()`` to execute two calls ``document.head.innerHTML`` and ``document.body.innerHTML``, so you will get the current page nodes.

Then, you can make a function that, when the user clicks on a button, it is called, like:

```python
async def save():
    html_head = await ui.run_javascript(document.head.innerHTML)
    html_body = await ui.run_javascript(document.body.innerHTML)
    # here you assemble the html again, and save it in a page.html
```
Remembering, I just thought, I'm not at home (work), so I can't test it, but give it a try, and let me know if it works.

**Reply by smojef** [reply]
I understand the complexities and limitations.  thanks,

**Comment by Shahin-rmz** [comment]
Nice discussion.
Is it possible to use [Jinja](https://jinja.palletsprojects.com/en/2.10.x/) in NiceGUI?
I want to make an invoice pdf generator as a practice project and want to know what is best solution to approach this problem with NiceGUI.

**Reply by falkoschindler** [reply]
Sure, you can use Jinja with NiceGUI. Actually, NiceGUI already uses Jinja templates.
But you probably won't interact with NiceGUI's core mechanics, but use Jinja in an isolated function or endpoint to generate some output, right?

**Reply by Shahin-rmz** [reply]
Awesome.
I do not exactly know how it works right now.
Will begin with the project and share with everyone in Discussions.

**Reply by Shahin-rmz** [reply]
Hi, I just started the project and want to play around to see how it works.
Could you please tell me why this results nothing?
:thinking: 

```python
name = "NiceName"
ui.html('this is {{ name }} using nicegui')

ui.run()
```

![image](https://user-images.githubusercontent.com/32805117/234209543-489f61cc-9d42-48dc-9188-65e938b5c883.png)

I think I **have to** use template right?

**Reply by phoskee** [reply]
```python
name = "NiceName"
ui.html(f'this is {name} using nicegui')

ui.run()
```

**Reply by Shahin-rmz** [reply]
> 
Nice, But I think this way I am using f-string and not jinja.
the `if statement` or `for loop` does not work.
```python
name = "NiceName"
ui.html(f'this is { name } using nicegui')


number = 30
ui.html(f'{% if==30 % } it is {number} {% else %} it is less')



ui.run()
```


and here is the log

```
 line 12
    (% if==30 % )
     ^
SyntaxError: f-string: invalid syntax

```

**Reply by falkoschindler** [reply]
@Shahin-rmz Sure, if you want to use Jinja, you'll need to tell Python to do so. The [Jinja2](https://pypi.org/project/Jinja2/) package is probably a good starting point. Or you use FastAPI's Jinja template support: https://fastapi.tiangolo.com/advanced/templates/

**Reply by Shahin-rmz** [reply]
This was my original code.
I have imported Jinja, but got this result.
![image](https://user-images.githubusercontent.com/32805117/234225606-d90d6f95-88f5-4227-b120-47b150ffb995.png)

```python
import jinja2
from nicegui import ui

ui.label('Hello NiceGUI!')
ui.button('BUTTON', on_click=lambda: ui.notify('button was pressed'))

name = "NiceName"
ui.html('this is {{ name }} using nicegui')


number = 30
ui.html('''
        {% if==30 % } </br>
        it is {number}</br>
        {% else %}</br>
        it is not''')



ui.run()
```

I think I should use FastAPI template support.
The inline method is not supported.
Or I am wrong :thinking: .

**Reply by falkoschindler** [reply]
Please have a look into the Jinja2 documentation or some tutorial. It doesn't work magically just by importing the module.

**Reply by Shahin-rmz** [reply]
@falkoschindler thanks for your hint.
I took a look inside the documentation and this is what I have end up.

```python
templates = Jinja2Templates(directory="templates")



@app.get("/users/{name}/{age}", response_class=HTMLResponse)
def submit(name: str, age: int, request: Request):
    return templates.TemplateResponse("greeting.html",{"request": request,
                                                   "name":name, "age":age})

name=ui.input(label='enter your name')
age=ui.number(label='enter the age',format='%d')


ui.button("submit a form",on_click=lambda:
          ui.open(f'/users/{name.value}/{age.value}'))

ui.run()
```
I should use `app.get` method instead of my last approach.
